File: util/anim/AnimUD.java
package common.util.anim;

import common.CommonStatic;
import common.system.VImg;
import common.system.fake.FakeImage;
import common.system.fake.FakeImage.Marker;
import common.system.files.FileData;
import common.system.files.VFile;

import java.util.ArrayList;
import java.util.List;

public class AnimUD extends AnimU<AnimUD.DefImgLoader> {

	private final String name;

	@Override
	public boolean cantLoadAll(ImageKeeper.AnimationType type) {
		return !loader.validate(type);
	}

	@Override
	public List<String> collectInvalidAnimation(ImageKeeper.AnimationType type) {
		return loader.collectInvalidAnimation(type);
	}

	static class DefImgLoader implements AnimU.ImageKeeper {

		private final String spath;
		private final VFile fnum, fedi, funi;
		private FileData dnum;
		private FakeImage num;
		private VImg edi, uni;

		private DefImgLoader(String path, String str, String sedi, String suni) {
			spath = path + str;
			fnum = VFile.get(spath + ".png");
			fedi = sedi == null ? null : VFile.get(path + sedi);
			funi = suni == null ? null : VFile.get(path + suni);
		}

		@Override
		public VImg getEdi() {
			if (edi != null && edi.getImg() != null && edi.getImg().bimg() != null && edi.getImg().isValid())
				return edi;

			return fedi == null ? null : (edi = new VImg(fedi).mark(Marker.EDI));
		}

		@Override
		public ImgCut getIC() {
			return ImgCut.newIns(spath + ".imgcut");
		}

		@Override
		public MaAnim[] getMA() {
			MaAnim[] ma;
			if (VFile.get(spath + ".maanim") != null) {
				ma = new MaAnim[] { MaAnim.newIns(spath + ".maanim") };
			} else {
				if (VFile.get(spath + "_zombie00.maanim") != null)
					ma = new MaAnim[7];
				else if (VFile.get(spath + "_entry.maanim") != null)
					ma = new MaAnim[5];
				else
					ma = new MaAnim[4];
				for (int i = 0; i < 4; i++)
					ma[i] = MaAnim.newIns(spath + "0" + i + ".maanim");
				if (ma.length == 5)
					ma[4] = MaAnim.newIns(spath + "_entry.maanim");
				if (ma.length == 7)
					for (int i = 0; i < 3; i++)
						ma[i + 4] = MaAnim.newIns(spath + "_zombie0" + i + ".maanim");
			}
			ma = filterValidAnims(ma);
			return ma;
		}

		@Override
		public MaModel getMM() {
			return MaModel.newIns(spath + ".mamodel");
		}

		@Override
		public FakeImage getNum() {
			if (num != null && num.bimg() != null && num.isValid())
				return num;

			FileData fd = dnum == null ? (dnum = fnum.getData()) : dnum;

			num = fd.getImg();

			return num;
		}

		@Override
		public VImg getUni() {
			if (uni != null && uni.getImg() != null && uni.getImg().bimg() != null && uni.getImg().isValid())
				return uni;

			if (funi == null) {
				return CommonStatic.getBCAssets().slot[0];
			} else {
				uni = new VImg(funi).mark(Marker.UNI);
				uni.setCut(CommonStatic.getBCAssets().unicut);

				return uni;
			}
		}

		@Override
		public void unload() {
			dnum = null;

			if (num != null) {
				num.unload();
				num = null;
			}

			if (edi != null) {
				edi.unload();
				edi = null;
			}

			if (uni != null) {
				uni.unload();
				uni = null;
			}
		}

		@Override
		public boolean validate(AnimationType type) {
			// This is for BC animations, if validate is false, just let program crash rather
			return true;
		}

		@Override
		public List<String> collectInvalidAnimation(AnimationType type) {
			// This is for BC animations, if it contains invalid animation, just let program crash rather
			return new ArrayList<>();
		}

		private MaAnim[] filterValidAnims(MaAnim[] original) {
			int end = 0;

			for (int i = 0; i < original.length; i++) {
				// walk/attack/wait/hb must be kept no matter what
				if ((original[i] != null && original[i].n != 0) || i < 4)
					end = i;
			}

			MaAnim[] fixed = new MaAnim[end + 1];

			System.arraycopy(original, 0, fixed, 0, end + 1);

			return fixed;
		}
	}

	public AnimUD(String path, String name, String edi, String uni) {
		super(path + name, new DefImgLoader(path, name, edi, uni));
		this.name = name;
	}

	@Override
	public String toString() {
		return name;
	}
}

File: util/anim/AnimUD.kt
package common.util.anim

import common.CommonStatic
import common.system.VImg
import common.system.fake.FakeImage
import common.system.fake.FakeImage.Marker
import common.system.files.FileData
import common.system.files.VFile
import common.util.anim.AnimUD.DefImgLoader
import common.util.anim.ImgCut
import common.util.anim.MaAnim
import common.util.anim.MaModel

class AnimUD(path: String, name: String, edi: String?, uni: String?) : AnimU<DefImgLoader?>(path + name, DefImgLoader(path, name, edi, uni)) {
    internal class DefImgLoader(path: String, str: String, sedi: String?, suni: String?) : ImageKeeper {
        private val spath: String
        private val fnum: VFile<*>
        private val fedi: VFile<*>?
        private val funi: VFile<*>?
        private var dnum: FileData? = null
        private var num: FakeImage? = null
        private var edi: VImg? = null
        private var uni: VImg? = null
        override fun getEdi(): VImg? {
            if (edi != null) return edi
            return if (fedi == null) null else VImg(fedi).mark(Marker.EDI).also { edi = it }
        }

        override fun getIC(): ImgCut {
            return ImgCut.Companion.newIns("$spath.imgcut")
        }

        override fun getMA(): Array<MaAnim?> {
            val ma: Array<MaAnim?>
            if (VFile.Companion.getFile(spath + "_zombie00.maanim") != null) ma = arrayOfNulls<MaAnim>(7) else if (VFile.Companion.getFile(spath + "_entry.maanim") != null) ma = arrayOfNulls<MaAnim>(5) else ma = arrayOfNulls<MaAnim>(4)
            for (i in 0..3) ma[i] = MaAnim.Companion.newIns(spath + "0" + i + ".maanim")
            if (ma.size == 5) ma[4] = MaAnim.Companion.newIns(spath + "_entry.maanim")
            if (ma.size == 7) for (i in 0..2) ma[i + 4] = MaAnim.Companion.newIns(spath + "_zombie0" + i + ".maanim")
            return ma
        }

        override fun getMM(): MaModel {
            return MaModel.Companion.newIns("$spath.mamodel")
        }

        override fun getNum(): FakeImage? {
            if (num != null) return num
            val fd = if (dnum == null) fnum.getData().also { dnum = it } else dnum!!
            num = fd.img
            return num
        }

        override fun getUni(): VImg? {
            if (uni != null) return uni
            return if (funi == null) CommonStatic.getBCAssets().slot.get(0) else VImg(funi).mark(Marker.UNI).also { uni = it }
        }

        override fun unload() {
            dnum = null
            if (num != null) {
                num.unload()
                num = null
            }
        }

        init {
            spath = path + str
            fnum = VFile.Companion.get("$spath.png")
            fedi = if (sedi == null) null else VFile.Companion.get(path + sedi)
            funi = if (suni == null) null else VFile.Companion.get(path + suni)
        }
    }
}
